### PR TITLE
`get_entry_point`: add proper deprecation pathway for old entry points 

### LINKED ISSUE #5202: Verdi: DataParamType cannot be used with deprecated entry points
- [x] [AiiDA Troubleshooting Documentation](https://aiida.readthedocs.io/projects/aiida-core/en/latest/intro/troubleshooting.html)
- [x] [AiiDA Users Forum](https://groups.google.com/forum/#!forum/aiidausers)

### Describe the bug

Using the ``DataParamType`` does not work with the now deprecated entry points like ``aiida.data:structure`` instead of ``aiida.data:core.structure``. It leads to errors like the following in the tests of ``aiida-fleur``.

```
________________________ ERROR collecting tests/cmdline/test_base_all_commands.py _________________________
../../aiida-core-fork/aiida/cmdline/params/types/identifier.py:62: in __init__
    entry_point = get_entry_point_from_string(entry_point_string)
../../aiida-core-fork/aiida/plugins/entry_point.py:183: in get_entry_point_from_string
    return get_entry_point(group, name)
../../aiida-core-fork/aiida/plugins/entry_point.py:265: in get_entry_point
    raise MissingEntryPointError(f"Entry point '{name}' not found in group '{group}'")
E   aiida.common.exceptions.MissingEntryPointError: Entry point 'structure' not found in group 'aiida.data'

During handling of the above exception, another exception occurred:
cmdline/test_base_all_commands.py:4: in <module>
    from aiida_fleur.cmdline import cmd_root
../aiida_fleur/cmdline/__init__.py:21: in <module>
    from .launch import cmd_launch
../aiida_fleur/cmdline/launch/__init__.py:16: in <module>
    from .launch import launch_inpgen
../aiida_fleur/cmdline/launch/launch.py:18: in <module>
    from ..util import options
../aiida_fleur/cmdline/util/options.py:31: in <module>
    type=types.DataParamType(sub_classes=('aiida.data:structure',)),
../../aiida-core-fork/aiida/cmdline/params/types/identifier.py:64: in __init__
    raise ValueError(f'{entry_point_string} is not a valid entry point string: {exception}')
E   ValueError: aiida.data:structure is not a valid entry point string: Entry point 'structure' not found in group 'aiida.data'
```
### Steps to reproduce

Create a command with an option of type ``DataParamType`` with a deprecated entry point

```
import click
from aiida.cmdline.types import DataParamType

@click.command('test')
@click.option('--data', type=DataParamType(sub_classes=('aiida.data:structure',)))
def test_command(identifier):
     pass

```

### Expected behavior

The old entry point should continue to work with a deprecation warning

### Your environment

- Operating system [e.g. Linux]: MacOS
- Python version [e.g. 3.7.1]: 3.9.7
- aiida-core version [e.g. 1.2.1]: develop

### PR DESCRIPTION
**Pull request by sphuber**
Fixes #5202 

All entry points that ship with aiida-core were deprecated for v2.0
as they were changed to be properly prefixed with core.. To make sure
the old entry points would still be automatically loaded, the factories
were updated to automatically catch them, print a deprecation warning,
and load the new entry point instead.

However, the factory was not the correct place to put this logic, since
the `get_entry_point` method, which the factories call and is the lowest
function in the stack that actually retrieves the entry point, can also
be called directly, circumventing the deprecation mechanic added to the
factories. This would result in the deprecated entry points raising an
exception when being loaded, for example in the parameter types of the
command line that have support for specific entry points, such as the
`IdentifierParamType`.

The solution is to move the deprecation mechanic from the factories to
the lowest layer of `get_entry_point`.

### CODE REVIEW

File: aiida/plugins/entry_point.py
Comment (chrisjsewell) [inline aiida/plugins/entry_point.py:285]: I would move this out of the function, so you don't have to recreate the dictionary on every function call. Probably a marginal gain, but don't want to slow down the CLI

**Review by ramirezfranciscof — COMMENTED** [review]
Thanks! Looks good to me in general, my only question is if it makes sense to have two separate functions doing this. I understand that maybe this is somehow "inherited" of the previous mechanism in which `warn_deprecated_entry_point` was called by many different factories with different lists of possible entry-points. But now that this is only required in `get_entry_point`, having that intermediate `convert_potentially_deprecated_entry_point` that only selects and passes the correct list of entry-points feels like bit forced separation of concerns. Could these be made a single function?

File: aiida/plugins/entry_point.py
Comment (sphuber) [reply aiida/plugins/entry_point.py]: Done

**Comment by sphuber** [comment]
Thanks @chrisjsewell and @ramirezfranciscof , addressed both your comments.

File: aiida/plugins/entry_point.py
Comment (chrisjsewell) [inline aiida/plugins/entry_point.py:302]: Is there a reason to not have these as top-level imports?

**Review by chrisjsewell — CHANGES_REQUESTED** [review]
one more thing cheers

File: aiida/plugins/entry_point.py
Comment (sphuber) [reply aiida/plugins/entry_point.py:302]: Because this function is the only place that uses them and when the function gets removed the obsolete imports are also automatically taken care over. Not a huge deal, but why not?

**Review by ramirezfranciscof — APPROVED** [review]
Thanks! Looks good to me! Feel free to merge after you finish your discussion with @chrisjsewell 

BTW this seems the best way to do multi-people review, right? I mean, each approves separately and it is understood the author would wait for approval from all before merging (this would be opposed to something like giving "approving" messages as comment but the last reviewer leaves theirs as "approve")

File: aiida/plugins/entry_point.py
Comment (chrisjsewell) [reply aiida/plugins/entry_point.py:302]: Meh, pylint will flag unused imports, so that's not an issue.
Yeh not a big deal, but that's the python coding standard that we use (outside reasons like CLI response and circular imports), whereas "single use imports should be internal to functions" is not.
Up to you

**Review by chrisjsewell — APPROVED** [review]

**Pull request merged**
